Request Circuit Breaker Bypass

**Elasticsearch version** (`bin/elasticsearch --version`):
Version: 7.2.0, Build: default/tar/508c38a/2019-06-20T15:54:18.811730Z, JVM: 12.0.1

**Plugins installed**:
No plugins install

**JVM version** (`java -version`):
java version "12" 2019-03-19
Java(TM) SE Runtime Environment (build 12+33)
Java HotSpot(TM) 64-Bit Server VM (build 12+33, mixed mode, sharing)

**OS version** (`uname -a` if on a Unix-like system):
Darwin hostname 18.6.0 Darwin Kernel Version 18.6.0: Thu Apr 25 23:16:27 PDT 2019; root:xnu-4903.261.4~2/RELEASE_X86_64 x86_64 i386 MacBookPro15,2 Darwin

**Description of the problem including expected versus actual behavior**:

A badly formatted request can lead to a double deallocation of request bytes in `ChildMemoryCircuitBreaker`. In particular:
-  a `BigArray` gets allocated in [BytesStreamOutput](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/common/io/stream/BytesStreamOutput.java#L62). It can get resized during the request, but nothing is wrong there.
- The breaker gets adjusted when [DefaultRestChannel](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/http/DefaultRestChannel.java#L133) calls `Releasables.close(toClose)`. The bytes are removed properly from the breaker. However, the arrays are still around.
- Due to the malformed request, `RestController` catches an exception and attempts to [send a response from the exception](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/rest/RestController.java#L176).
- Building the response calls `BytesStreamOutput.reset()`, which calls resize on an array that was released earlier. This can lead to removing bytes from the circuit Breaker, even though they were already removed.
- At the end of the request, `ChildMemoryCircuitBreaker.used` becomes a negative value, and persists in subsequent requests. By sending that malformed request multiple times, a client could set the ChildMemoryCircuitBreaker arbitrary low, which could lead to bypassing the circuit breaking. This would lead to OutOfMemory exceptions when combined with memory-intensive queries. Since the CircuitBreaker counter is 64-bit, I do not believe an attacker could realistically underflow it to prevent further allocations.

**Steps to reproduce**:

Please include a *minimal* but *complete* recreation of the problem, including
(e.g.) index creation, mappings, settings, query etc.  The easier you make for
us to reproduce it, the more likely that somebody will take the time to look at it.

 1. Start a fresh elastic 
 2. Send malformed request: curl `https://gist.githubusercontent.com/alpire/9a0ad1dae076ee34eba74f72eac15d83/raw/51bad3aac237cf6fc58a957a10182a9bc29924e1/es-input.hexdump | xxd -r | nc localhost 9200`
 3. See exception in output, or look at the  negative value of `ChildMemoryCircuitBreaker.used` in a debugger in subsequent requests.

**Provide logs (if relevant)**:

With assertions disabled, ES fails with a `NullPointerException` after adjusting the breaker:
```
[2019-07-01T20:37:50,902][ERROR][o.e.r.RestController     ] [...] failed to send failure response for uri [...]
java.lang.NullPointerException: null
        at org.elasticsearch.common.util.AbstractBigArray.releasePage(AbstractBigArray.java:153) ~[elasticsearch-7.2.0.jar:7.2.0]
        at org.elasticsearch.common.util.BigByteArray.resize(BigByteArray.java:147) ~[elasticsearch-7.2.0.jar:7.2.0]
        at org.elasticsearch.common.util.BigByteArray.resize(BigByteArray.java:147) ~[elasticsearch-7.2.0.jar:7.2.0]
        at org.elasticsearch.common.util.BigArrays.resizeInPlace(BigArrays.java:440) ~[elasticsearch-7.2.0.jar:7.2.0]
        at org.elasticsearch.common.util.BigArrays.resize(BigArrays.java:487) ~[elasticsearch-7.2.0.jar:7.2.0]
        at org.elasticsearch.common.io.stream.BytesStreamOutput.reset(BytesStreamOutput.java:103) ~[elasticsearch-7.2.0.jar:7.2.0]
        at org.elasticsearch.common.io.stream.ReleasableBytesStreamOutput.reset(ReleasableBytesStreamOutput.java:80) ~[elasticsearch-7.2.0.jar:7.2.0]
        at org.elasticsearch.rest.AbstractRestChannel.bytesOutput(AbstractRestChannel.java:130) ~[elasticsearch-7.2.0.jar:7.2.0]
        at org.elasticsearch.rest.AbstractRestChannel.newBuilder(AbstractRestChannel.java:109) ~[elasticsearch-7.2.0.jar:7.2.0]
        at org.elasticsearch.rest.AbstractRestChannel.newErrorBuilder(AbstractRestChannel.java:75) ~[elasticsearch-7.2.0.jar:7.2.0]
        at org.elasticsearch.rest.BytesRestResponse.build(BytesRestResponse.java:137) ~[elasticsearch-7.2.0.jar:7.2.0]
        at org.elasticsearch.rest.BytesRestResponse.<init>(BytesRestResponse.java:96) ~[elasticsearch-7.2.0.jar:7.2.0]
        at org.elasticsearch.rest.BytesRestResponse.<init>(BytesRestResponse.java:91) ~[elasticsearch-7.2.0.jar:7.2.0]
        at org.elasticsearch.rest.RestController.dispatchRequest(RestController.java:177) [elasticsearch-7.2.0.jar:7.2.0]
        at org.elasticsearch.http.AbstractHttpServerTransport.dispatchRequest(AbstractHttpServerTransport.java:320) [elasticsearch-7.2.0.jar:7.2.0]
        at org.elasticsearch.http.AbstractHttpServerTransport.handleIncomingRequest(AbstractHttpServerTransport.java:370) [elasticsearch-7.2.0.jar:7.2.0]
        at org.elasticsearch.http.AbstractHttpServerTransport.incomingRequest(AbstractHttpServerTransport.java:299) [elasticsearch-7.2.0.jar:7.2.0]
        at org.elasticsearch.http.netty4.Netty4HttpRequestHandler.channelRead0(Netty4HttpRequestHandler.java:66) [transport-netty4-client-7.2.0.jar:7.2.0]
        at org.elasticsearch.http.netty4.Netty4HttpRequestHandler.channelRead0(Netty4HttpRequestHandler.java:31) [transport-netty4-client-7.2.0.jar:7.2.0]
        at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:105) [netty-transport-4.1.35.Final.jar:4.1.35.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374) [netty-transport-4.1.35.Final.jar:4.1.35.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:360) [netty-transport-4.1.35.Final.jar:4.1.35.Final]
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:352) [netty-transport-4.1.35.Final.jar:4.1.35.Final]
        at org.elasticsearch.http.netty4.Netty4HttpPipeliningHandler.channelRead(Netty4HttpPipeliningHandler.java:58) [transport-netty4-client-7.2.0.jar:7.2.0]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374) [netty-transport-4.1.35.Final.jar:4.1.35.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:360) [netty-transport-4.1.35.Final.jar:4.1.35.Final]
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:352) [netty-transport-4.1.35.Final.jar:4.1.35.Final]
        at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:102) [netty-codec-4.1.35.Final.jar:4.1.35.Final]
        at io.netty.handler.codec.MessageToMessageCodec.channelRead(MessageToMessageCodec.java:111) [netty-codec-4.1.35.Final.jar:4.1.35.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374) [netty-transport-4.1.35.Final.jar:4.1.35.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:360) [netty-transport-4.1.35.Final.jar:4.1.35.Final]
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:352) [netty-transport-4.1.35.Final.jar:4.1.35.Final]
        at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:102) [netty-codec-4.1.35.Final.jar:4.1.35.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374) [netty-transport-4.1.35.Final.jar:4.1.35.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:360) [netty-transport-4.1.35.Final.jar:4.1.35.Final]
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:352) [netty-transport-4.1.35.Final.jar:4.1.35.Final]
        at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:102) [netty-codec-4.1.35.Final.jar:4.1.35.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374) [netty-transport-4.1.35.Final.jar:4.1.35.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:360) [netty-transport-4.1.35.Final.jar:4.1.35.Final]
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:352) [netty-transport-4.1.35.Final.jar:4.1.35.Final]
        at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:323) [netty-codec-4.1.35.Final.jar:4.1.35.Final]
        at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:297) [netty-codec-4.1.35.Final.jar:4.1.35.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374) [netty-transport-4.1.35.Final.jar:4.1.35.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:360) [netty-transport-4.1.35.Final.jar:4.1.35.Final]
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:352) [netty-transport-4.1.35.Final.jar:4.1.35.Final]
        at io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:287) [netty-handler-4.1.35.Final.jar:4.1.35.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374) [netty-transport-4.1.35.Final.jar:4.1.35.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:360) [netty-transport-4.1.35.Final.jar:4.1.35.Final]
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:352) [netty-transport-4.1.35.Final.jar:4.1.35.Final]
        at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1408) [netty-transport-4.1.35.Final.jar:4.1.35.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374) [netty-transport-4.1.35.Final.jar:4.1.35.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:360) [netty-transport-4.1.35.Final.jar:4.1.35.Final]
        at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:930) [netty-transport-4.1.35.Final.jar:4.1.35.Final]
        at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163) [netty-transport-4.1.35.Final.jar:4.1.35.Final]
        at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:682) [netty-transport-4.1.35.Final.jar:4.1.35.Final]
        at io.netty.channel.nio.NioEventLoop.processSelectedKeysPlain(NioEventLoop.java:582) [netty-transport-4.1.35.Final.jar:4.1.35.Final]
        at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:536) [netty-transport-4.1.35.Final.jar:4.1.35.Final]
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:496) [netty-transport-4.1.35.Final.jar:4.1.35.Final]
        at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:906) [netty-common-4.1.35.Final.jar:4.1.35.Final]
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) [netty-common-4.1.35.Final.jar:4.1.35.Final]
        at java.lang.Thread.run(Thread.java:835) [?:?]
        Suppressed: java.lang.IllegalArgumentException: Unexpected http protocol version: ETTP/1.1
                at org.elasticsearch.http.netty4.Netty4HttpRequest.protocolVersion(Netty4HttpRequest.java:136) ~[?:?]
                at org.elasticsearch.http.DefaultRestChannel.isHttp10(DefaultRestChannel.java:172) ~[elasticsearch-7.2.0.jar:7.2.0]
                at org.elasticsearch.http.DefaultRestChannel.isCloseConnection(DefaultRestChannel.java:166) ~[elasticsearch-7.2.0.jar:7.2.0]
                at org.elasticsearch.http.DefaultRestChannel.sendResponse(DefaultRestChannel.java:118) ~[elasticsearch-7.2.0.jar:7.2.0]
                at org.elasticsearch.rest.RestController.handleUnsupportedHttpMethod(RestController.java:373) ~[elasticsearch-7.2.0.jar:7.2.0]
                ...
```

With assertions enabled, you can clearly see that the used bytes become negative:
```
[2019-07-01T21:18:41,734][WARN ][o.e.h.AbstractHttpServerTransport] [...] caught exception while handling client http traffic, closing connection Netty4HttpChannel{localAddress=/0:0:0:0:0:0:0:1:9200, remoteAddress=/0:0:0:0:0:0:0:1:55682}
...
Caused by: java.lang.AssertionError: Used bytes: [-16384] must be >= 0
        at org.elasticsearch.common.breaker.ChildMemoryCircuitBreaker.addWithoutBreaking(ChildMemoryCircuitBreaker.java:194) ~[elasticsearch-7.2.0.jar:7.2.0]
        at org.elasticsearch.common.util.BigArrays.adjustBreaker(BigArrays.java:415) ~[elasticsearch-7.2.0.jar:7.2.0]
        at org.elasticsearch.common.util.BigArrays.resizeInPlace(BigArrays.java:439) ~[elasticsearch-7.2.0.jar:7.2.0]
        at org.elasticsearch.common.util.BigArrays.resize(BigArrays.java:487) ~[elasticsearch-7.2.0.jar:7.2.0]
        at org.elasticsearch.common.io.stream.BytesStreamOutput.reset(BytesStreamOutput.java:103) ~[elasticsearch-7.2.0.jar:7.2.0]
        at org.elasticsearch.common.io.stream.ReleasableBytesStreamOutput.reset(ReleasableBytesStreamOutput.java:80) ~[elasticsearch-7.2.0.jar:7.2.0]
        at org.elasticsearch.rest.AbstractRestChannel.bytesOutput(AbstractRestChannel.java:130) ~[elasticsearch-7.2.0.jar:7.2.0]
        at org.elasticsearch.rest.AbstractRestChannel.newBuilder(AbstractRestChannel.java:109) ~[elasticsearch-7.2.0.jar:7.2.0]
        at org.elasticsearch.rest.AbstractRestChannel.newErrorBuilder(AbstractRestChannel.java:75) ~[elasticsearch-7.2.0.jar:7.2.0]
        at org.elasticsearch.rest.BytesRestResponse.build(BytesRestResponse.java:137) ~[elasticsearch-7.2.0.jar:7.2.0]
        at org.elasticsearch.rest.BytesRestResponse.<init>(BytesRestResponse.java:96) ~[elasticsearch-7.2.0.jar:7.2.0]
        at org.elasticsearch.rest.BytesRestResponse.<init>(BytesRestResponse.java:91) ~[elasticsearch-7.2.0.jar:7.2.0]
        at org.elasticsearch.rest.RestController.dispatchRequest(RestController.java:177) ~[elasticsearch-7.2.0.jar:7.2.0]
        at org.elasticsearch.http.AbstractHttpServerTransport.dispatchRequest(AbstractHttpServerTransport.java:320) ~[elasticsearch-7.2.0.jar:7.2.0]
        at org.elasticsearch.http.AbstractHttpServerTransport.handleIncomingRequest(AbstractHttpServerTransport.java:370) ~[elasticsearch-7.2.0.jar:7.2.0]
        at org.elasticsearch.http.AbstractHttpServerTransport.incomingRequest(AbstractHttpServerTransport.java:299) ~[elasticsearch-7.2.0.jar:7.2.0]
        at org.elasticsearch.http.netty4.Netty4HttpRequestHandler.channelRead0(Netty4HttpRequestHandler.java:66) ~[?:?]
        at org.elasticsearch.http.netty4.Netty4HttpRequestHandler.channelRead0(Netty4HttpRequestHandler.java:31) ~[?:?]
        at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:105) ~[?:?]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374) ~[?:?]
        ... 40 more
```

--- a/server/src/main/java/org/elasticsearch/http/DefaultRestChannel.java
+++ b/server/src/main/java/org/elasticsearch/http/DefaultRestChannel.java
@@ -168,12 +168,13 @@ public class DefaultRestChannel extends AbstractRestChannel implements RestChann
 
     // Determine if the request connection should be closed on completion.
     private boolean isCloseConnection() {
-        final boolean http10 = isHttp10();
-        return CLOSE.equalsIgnoreCase(request.header(CONNECTION)) || (http10 && !KEEP_ALIVE.equalsIgnoreCase(request.header(CONNECTION)));
-    }
-
-    // Determine if the request protocol version is HTTP 1.0
-    private boolean isHttp10() {
-        return request.getHttpRequest().protocolVersion() == HttpRequest.HttpVersion.HTTP_1_0;
+        try {
+            final boolean http10 = request.getHttpRequest().protocolVersion() == HttpRequest.HttpVersion.HTTP_1_0;
+            return CLOSE.equalsIgnoreCase(request.header(CONNECTION))
+                || (http10 && !KEEP_ALIVE.equalsIgnoreCase(request.header(CONNECTION)));
+        } catch (Exception e) {
+            // In case we fail to parse the http protocol version out of the request we always close the connection
+            return true;
+        }
     }
 }

--- a/server/src/test/java/org/elasticsearch/http/DefaultRestChannelTests.java
+++ b/server/src/test/java/org/elasticsearch/http/DefaultRestChannelTests.java
@@ -57,6 +57,7 @@ import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.Objects;
+import java.util.function.Supplier;
 
 import static org.hamcrest.Matchers.equalTo;
 import static org.hamcrest.Matchers.hasItem;
@@ -272,8 +273,13 @@ public class DefaultRestChannelTests extends ESTestCase {
     public void testConnectionClose() throws Exception {
         final Settings settings = Settings.builder().build();
         final HttpRequest httpRequest;
-        final boolean close = randomBoolean();
-        if (randomBoolean()) {
+        final boolean brokenRequest = randomBoolean();
+        final boolean close = brokenRequest || randomBoolean();
+        if (brokenRequest) {
+            httpRequest = new TestRequest(() -> {
+                throw new IllegalArgumentException("Can't parse HTTP version");
+            }, RestRequest.Method.GET, "/");
+        } else if (randomBoolean()) {
             httpRequest = new TestRequest(HttpRequest.HttpVersion.HTTP_1_1, RestRequest.Method.GET, "/");
             if (close) {
                 httpRequest.getHeaders().put(DefaultRestChannel.CONNECTION, Collections.singletonList(DefaultRestChannel.CLOSE));
@@ -399,16 +405,19 @@ public class DefaultRestChannelTests extends ESTestCase {
 
     private static class TestRequest implements HttpRequest {
 
-        private final HttpVersion version;
+        private final Supplier<HttpVersion> version;
         private final RestRequest.Method method;
         private final String uri;
         private HashMap<String, List<String>> headers = new HashMap<>();
 
-        private TestRequest(HttpVersion version, RestRequest.Method method, String uri) {
-
-            this.version = version;
+        private TestRequest(Supplier<HttpVersion> versionSupplier, RestRequest.Method method, String uri) {
+            this.version = versionSupplier;
             this.method = method;
             this.uri = uri;
+        }
+
+        private TestRequest(HttpVersion version, RestRequest.Method method, String uri) {
+            this(() -> version, method, uri);
         }
 
         @Override
@@ -438,7 +447,7 @@ public class DefaultRestChannelTests extends ESTestCase {
 
         @Override
         public HttpVersion protocolVersion() {
-            return version;
+            return version.get();
         }
 
         @Override

